datafusion/core/src/physical_plan/repartition/sort_preserving_repartition.rs

```diff
@@ -0,0 +1,357 @@
+            .map(|receiver| {
+                Box::pin(PerPartitionStream {
+                    schema: self.schema(),
+                    receiver,
+                    drop_helper: Arc::clone(&state.abort_helper),
+                    reservation: reservation.clone(),
+                }) as SendableRecordBatchStream
+            })
+            .collect::<Vec<_>>();
+        // Note that receiver size (`rx.len()`) and `num_input_partitions` are same.
+
+        // Get existing ordering:
+        let sort_exprs = self.input.output_ordering().unwrap_or(&[]);
+        // Merge streams (while preserving ordering) coming from input partitions to this partition:
+        streaming_merge(
+            input_streams,
+            self.schema(),
+            sort_exprs,
+            BaselineMetrics::new(&self.metrics, partition),
+            context.session_config().batch_size(),
+        )
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+
+    fn fmt_as(
+        &self,
+        t: DisplayFormatType,
+        f: &mut std::fmt::Formatter,
+    ) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                write!(
+                    f,
+                    "SortPreservingRepartitionExec: partitioning={:?}, input_partitions={}",
+                    self.partitioning,
+                    self.input.output_partitioning().partition_count()
+                )
+            }
+        }
+    }
+
+    fn statistics(&self) -> Statistics {
+        self.input.statistics()
+    }
+}
+
+impl SortPreservingRepartitionExec {
+    /// Create a new `SortPreservingRepartitionExec`.
+    pub fn try_new(
+        input: Arc<dyn ExecutionPlan>,
+        partitioning: Partitioning,
+    ) -> Result<Self> {
+        Ok(SortPreservingRepartitionExec {
+            input,
+            partitioning,
+            state: Arc::new(Mutex::new(SortPreservingRepartitionExecState {
+                channels: HashMap::new(),
+                abort_helper: Arc::new(AbortOnDropMany::<()>(vec![])),
+            })),
+            metrics: ExecutionPlanMetricsSet::new(),
+        })
+    }
+}
+
+/// This struct converts a receiver to a stream.
+/// Receiver receives data on an SPSC channel.
+struct PerPartitionStream {
+    /// Schema wrapped by Arc
+    schema: SchemaRef,
+
+    /// channel containing the repartitioned batches
+    receiver: DistributionReceiver<MaybeBatch>,
+
+    /// Handle to ensure background tasks are killed when no longer needed.
+    #[allow(dead_code)]
+    drop_helper: Arc<AbortOnDropMany<()>>,
+
+    /// Memory reservation.
+    reservation: SharedMemoryReservation,
+}
+
+impl Stream for PerPartitionStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        match self.receiver.recv().poll_unpin(cx) {
```

Comment:
You could use the [`ready!` macro ](https://docs.rs/futures/0.3.28/futures/macro.ready.html)here if you wanted and save some repetition

```rust
        match ready!(self.receiver.recv().poll_unpin(cx)) {
            Some(Some(v)) => {
                if let Ok(batch) = &v {
                    self.reservation
                        .lock()
                        .shrink(batch.get_array_memory_size());
                }
                Poll::Ready(Some(v))
            }
            Some(None) => {
                // Input partition has finished sending batches
                Poll::Ready(None)
            }
            None => Poll::Ready(None),
        }
```